File: generated_code/cpp/solution_1093_3.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* recoverFromPreorder(string traversal) {
        // Stack to keep track of the path from the root to the current node.
        // The size of the stack matches the depth of the next node to be inserted.
        vector<TreeNode*> path;
        int i = 0;
        int n = traversal.length();
        
        while (i < n) {
            // Count the number of dashes to determine the depth of the current node
            int depth = 0;
            while (i < n && traversal[i] == '-') {
                depth++;
                i++;
            }
            
            // Extract the value of the current node
            int val = 0;
            while (i < n && traversal[i] != '-') {
                val = val * 10 + (traversal[i] - '0');
                i++;
            }
            
            TreeNode* node = new TreeNode(val);
            
            // If the stack size is greater than the current depth, it means we have 
            // finished processing the subtree of the deeper nodes, so we pop them.
            while (path.size() > depth) {
                path.pop_back();
            }
            
            // Attach the node to its parent if the stack is not empty
            if (!path.empty()) {
                if (!path.back()->left) {
                    path.back()->left = node;
                } else {
                    path.back()->right = node;
                }
            }
            
            // Push the current node onto the path stack
            path.push_back(node);
        }
        
        // The first element in the path is always the root of the tree
        return path[0];
    }
};

File: generated_code/python/solution_1093_3.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def recoverFromPreorder(self, traversal: str) -> Optional[TreeNode]:
        stack = []
        i = 0
        n = len(traversal)
        
        while i < n:
            # 1. Count the number of dashes to determine the depth of the current node
            depth = 0
            while i < n and traversal[i] == '-':
                depth += 1
                i += 1
            
            # 2. Extract the integer value of the current node
            val = 0
            while i < n and traversal[i].isdigit():
                val = val * 10 + int(traversal[i])
                i += 1
            
            node = TreeNode(val)
            
            # 3. Maintain the stack so its size matches the required depth.
            # If the stack size is greater than the depth, it means we are backtracking
            # up the tree, so we pop elements that are no longer parents of the current node.
            while len(stack) > depth:
                stack.pop()
            
            # 4. Attach the node to its parent.
            # If the stack is not empty, the top of the stack is the parent.
            if stack:
                parent = stack[-1]
                # Since we always try to populate the left child first, if left is free,
                # we assign it there. Otherwise, it goes to the right child.
                if not parent.left:
                    parent.left = node
                else:
                    parent.right = node
            
            # 5. Push the current node onto the stack as a potential parent for future nodes
            stack.append(node)
            
        # The first element in the stack will always be the root of the tree
        return stack[0] if stack else None